server/routers/viewer.tsx

```diff
@@ -91,6 +94,87 @@ export const viewerRouter = createProtectedRouter()
       return bookings;
     },
   })
+  .query("integrations", {
+    async resolve({ ctx }) {
+      const { user } = ctx;
+      const { credentials } = user;
+      const integrations = getIntegrations(credentials);
+
+      function countActive(items: { credentials: unknown[] }[]) {
+        return items.reduce((acc, item) => acc + item.credentials.length, 0);
+      }
+      const conferencing = integrations.flatMap((item) => (item.variant === "conferencing" ? [item] : []));
+      const payment = integrations.flatMap((item) => (item.variant === "payment" ? [item] : []));
+      const calendar = integrations.flatMap((item) => (item.variant === "calendar" ? [item] : []));
+
+      // get user's credentials + their connected integrations
+      const calendarCredentials = user.credentials
+        .filter((credential) => credential.type.endsWith("_calendar"))
+        .flatMap((credential) => {
+          const integration = ALL_INTEGRATIONS.find((integration) => integration.type === credential.type);
+
+          const adapter = getCalendarAdapterOrNull({
+            ...credential,
+            userId: user.id,
+          });
+          return integration && adapter && integration.variant === "calendar"
+            ? [{ integration, credential, adapter }]
+            : [];
+        });
+
+      // get all the connected integrations' calendars (from third party)
+      const connectedCalendars = await Promise.all(
+        calendarCredentials.map(async (item) => {
+          const { adapter, integration, credential } = item;
+          try {
+            const _calendars = await adapter.listCalendars();
+            const calendars = _calendars.map((cal) => ({
+              ...cal,
+              isSelected: !!user.selectedCalendars.find((selected) => selected.externalId === cal.externalId),
+            }));
+            const primary = calendars.find((item) => item.primary);
+            if (!primary) {
+              return {
+                integration,
+                error: {
+                  message: "No primary calendar found",
+                },
+              };
+            }
+            return {
+              integration,
+              credentialId: credential.id,
+              primary,
+              calendars,
+            };
+          } catch (_error) {
+            const error = getErrorFromUnknown(_error);
+            return {
+              integration,
+              error: {
+                message: error.message,
+              },
+            };
+          }
+        })
+      );
```

Comment:
I'm sorry - this part is a mess - a lot [due to this](https://calendso.slack.com/archives/C022UB8FMGT/p1633687094110200)